functions-continued/variable-scope.md

```diff
@@ -0,0 +1,339 @@
+    print("It is now time to declare each lunch item, slowly, and dramatically.")
+    for lunch_item in lunch_items:
+        print("Today we will eat......")
+        print("a.....")
+        print(lunch_item)
+    print("I hope you're hungry!")
+
+packed_lunchbox_contents = ["sandwich", "salad", "chips", "appy juice"]
+announce_lunch_menu(packed_lunchbox_contents)
+```
+
+## Summary
+
+Variable scope is the area of a program where a variable can be accessed.
+
+When we access variables that are within scope, we can read them, manipulate them, re-assign them, and do all the good things that variables can do.
+
+When we access variables out of scope, we get errors. We often get errors like `NameError`.
+
+We will often make **local variables** within functions, that are scoped to that function.
+
+## Check for Understanding
+
+Finish this sentence: Variable scope is...
+
+* the area of a program where a variable can be defined for the first time
+* the area of a program where a variable can be accessed
+* the area of a program where a variable can be released from memory
+* the area of a program where a variable can be renamed
+
+
+Read this code, and assume `cut`, `cook`, and `plate` are all valid, defined functions.
+
+```python
+def prepare_dinner(ingredients):
+    chopped_food = cut(ingredients)
+    cooked_food = cook(chopped_food)
+    plated_food = plate(cooked_food)
+    return plated_food
+
+prepare_dinner(["Jackfruit", "Onion", "Mango", "Avocado", "Ginger", "Cucumber", "Lime", "Jalapeño"])
+```
+Variables that have can be read inside of the function are...
+
+* `ingredients`
+* `chopped_food`, `cooked_food`, `plated_food`
+* `ingredients`, `chopped_food`, `cooked_food`, `plated_food`
+* `ingredients`, `plated_food`
+
+
+
+Read this code, and assume `cut`, `cook`, and `plate` are all valid, defined functions.
+
+```python
+def prepare_dinner(ingredients):
+    chopped_food = cut(ingredients)
+    cooked_food = cook(chopped_food)
+    plated_food = plate(cooked_food)
+    return plated_food
+
+dinner_recipe = ["Jackfruit", "Onion", "Mango", "Avocado", "Ginger", "Cucumber", "Lime", "Jalapeño"]
+prepare_dinner(dinner_recipe)
+```
+
+Which of these describe the variables (and their scope) inside of this code?
+
+- `dinner_recipe`, `chopped_food`, `cooked_food`, and `plated_food` are all local variables. They can all be accessed inside of the function.
+- `dinner_recipe`, `ingredients`, `chopped_food`, `cooked_food`, and `plated_food` are all local variables. They can all be accessed inside of the function.
+- `dinner_recipe` is a global variable. `ingredients` is a parameter. `chopped_food`, `cooked_food`, and `plated_food` are local variables. Only the parameters and local variables can be accessed inside of the function.
+- `dinner_recipe` is a global variable. `ingredients` is a parameter. `chopped_food`, `cooked_food`, and `plated_food` are local variables. They can all be accessed inside of the function.
+
```

Comment:
```suggestion
```suggestion
### !challenge

* type: multiple-choice
* id: 45954838-a89d-4b65-a5db-e6e52aae3ed7
* title: Variable Scope

##### !question

Finish this sentence: Variable scope is...

##### !end-question

##### !options

* the area of a program where a variable can be defined for the first time
* the area of a program where a variable can be accessed
* the area of a program where a variable can be released from memory
* the area of a program where a variable can be renamed

##### !end-options

##### !answer

* the area of a program where a variable can be accessed

##### !end-answer

### !end-challenge

### !challenge

* type: multiple-choice
* id: e3584e35-5c7d-4e46-a210-a1a0b697e8c0
* title: Local Variables

##### !question

Read this code, and assume `cut`, `cook`, and `plate` are all valid, defined functions.

```python
def prepare_dinner(ingredients):
    chopped_food = cut(ingredients)
    cooked_food = cook(chopped_food)
    plated_food = plate(cooked_food)
    return plated_food

prepare_dinner(["Jackfruit", "Onion", "Mango", "Avocado", "Ginger", "Cucumber", "Lime", "Jalapeño"])
```
Variables that have can be read inside of the function are...
##### !end-question

##### !options

* `ingredients`
* `chopped_food`, `cooked_food`, `plated_food`
* `ingredients`, `chopped_food`, `cooked_food`, `plated_food`
* `ingredients`, `plated_food`


##### !end-options

##### !answer

* `ingredients`, `chopped_food`, `cooked_food`, `plated_food`

##### !end-answer

### !end-challenge


### !challenge

* type: multiple-choice
* id: e9704a8e-e835-44c3-8bf0-57433d7654e9
* title: Describe These Variables

##### !question

Read this code, and assume `cut`, `cook`, and `plate` are all valid, defined functions.

```python
def prepare_dinner(ingredients):
    chopped_food = cut(ingredients)
    cooked_food = cook(chopped_food)
    plated_food = plate(cooked_food)
    return plated_food

dinner_recipe = ["Jackfruit", "Onion", "Mango", "Avocado", "Ginger", "Cucumber", "Lime", "Jalapeño"]
prepare_dinner(dinner_recipe)
```

Which of these describe the variables (and their scope) inside of this code?

##### !end-question

##### !options

* `dinner_recipe`, `chopped_food`, `cooked_food`, and `plated_food` are all local variables. They can all be accessed inside of the function.
* `dinner_recipe`, `ingredients`, `chopped_food`, `cooked_food`, and `plated_food` are all local variables. They can all be accessed inside of the function.
* `dinner_recipe` is a global variable. `ingredients` is a parameter. `chopped_food`, `cooked_food`, and `plated_food` are local variables. Only the parameters and local variables can be accessed inside of the function.
* `dinner_recipe` is a global variable. `ingredients` is a parameter. `chopped_food`, `cooked_food`, and `plated_food` are local variables. They can all be accessed inside of the function.

##### !end-options

##### !answer

* `dinner_recipe` is a global variable. `ingredients` is a parameter. `chopped_food`, `cooked_food`, and `plated_food` are local variables. They can all be accessed inside of the function.

##### !end-answer

### !end-challenge
```
```